src/manifest_parser.cc

```diff
@@ -152,6 +152,10 @@ bool ManifestParser::ParseRule(string* err) {
       return false;
 
     if (Rule::IsReservedBinding(key)) {
+      for (auto it = value.parsed_.begin(); it != value.parsed_.end(); ++it) {
+        it->first = it->first.PersistOutsideArena();
+      }
+      value.single_token_ = value.single_token_.PersistOutsideArena();
```

Comment:
Correct me if I'm wrong, but this looks like a memory  leak, since StringPiece is designed to be a non-owning view over string content, the result of PersistOutsideArena() will never be released. This method is probably confusing: returning an _owning_ StringPiece value is surprising and error-prone, using StringPiece::AsString() is much better because the lifetime of an std::string is easy to understand.

Reply:
I'm not entirely sure if I understand. Everything allocated on an arena is, in a sense, a memory leak; it's not freed until the arena goes out of scope. But at that point, it really is.

Are you saying that the lifetime of the arena here is unreasonably long?

Reply:
Wait, I'm probably confused. I'll need to reacquaint myself with my own code, it has been so long now. :-)